book/src/read-only-read-write-accounts.md

```diff
@@ -0,0 +1,135 @@
+# Read-Only/Read-Write Transaction Accounts
+
+This design covers handling of read-only and read-write accounts in the
+[runtime](runtime.md).  Accounts already distinguish themselves as read-only or
+read-write based on the program id specified by the transaction's instruction.
+Programs must treat accounts that are not owned by them as read-only.
+
+To identify read-only accounts by program id would require the Account to be
+fetched and loaded from disk.  This operation is expensive and while it is
+occurring the runtime would have to reject any transactions referencing the same
+account.
+
+The proposal introduces a `read_only_accounts` field to the Transaction
+structure and removes the `program_ids` dedicated vector for program accounts.
+
+This design doesn't change the runtime transaction processing rules.
+Programs still can't write or spend accounts that they do not own.  But it
+allows the runtime to optimistically take the right lock for each account
+specified in the transaction before loading the accounts from storage.
+
+Accounts selected as read-write by the transaction can still be treated as
+read-only by the instructions.
+
+## Runtime handling
+
+Read-only accounts have the following properties:
+
+* Can be deposited into.  Deposits can be implemented as a simple `atomic_add`.
```

Comment:
That's why I've always refer to these accounts as "credit-only", not "read-only". Want to do the same in this doc?

Reply:
I think how memory is treated is the more significant part.  RC, read-credit? RWW, read-withdraw-write?

Reply:
How about credit-only and credit-debit accounts?  How memory is treaded is only the more significant part of this particular proposal. How currency is handled might be more significant to users. In any case, we should always aim to CIWII, and because the runtime allows lamports to increase, it's imprecise to call the account read-only.

Reply:
Funny how even your title would become short and sweet if you used "credit-only". Instead of "Add proposal for Read-Only accounts that allow credits" it'd reduce to "Add proposal for credit-only accounts".

Reply:
@garious 
* ROC, read-only-credit
* RWW, read-write-withdraw

Reply:
That's unnecessarily verbose. These are attributes of things called "accounts". You could say that the userdata of a credit-only account is read-only, and the userdata of a credit-debit account is read-write. Introducing acronyms here is not necessary or helpful.

Reply:
@garious it helps to identify the accounts referenced in instructions as RO or RW, most of the time the token's are not important because the data is what is observed.  Do you want to identify them as C, or W?

Reply:
R and RW seems fine there. You're referring to the userdata.

Reply:
@garious yea, so what is the convention we want to use?  I'd like to have a standard one that is not to verbose.  CO/CD?

Reply:
Sure, CO/CD. Let's just be consistent in case we want to globally search and replace it later.